Mathlib/LinearAlgebra/Alternating.lean

```diff
@@ -0,0 +1,1262 @@
+
+/-!
+# Alternating Maps
+
+We construct the bundled function `AlternatingMap`, which extends `MultilinearMap` with all the
+arguments of the same type.
+
+## Main definitions
+* `AlternatingMap R M N ι` is the space of `R`-linear alternating maps from `ι → M` to `N`.
+* `f.map_eq_zero_of_eq` expresses that `f` is zero when two inputs are equal.
+* `f.map_swap` expresses that `f` is negated when two inputs are swapped.
+* `f.map_perm` expresses how `f` varies by a sign change under a permutation of its inputs.
+* An `AddCommMonoid`, `AddCommGroup`, and `Module` structure over `AlternatingMap`s that
+  matches the definitions over `MultilinearMap`s.
+* `MultilinearMap.domDomCongr`, for permutating the elements within a family.
+* `MultilinearMap.alternatization`, which makes an alternating map out of a non-alternating one.
+* `AlternatingMap.domCoprod`, which behaves as a product between two alternating maps.
+* `AlternatingMap.curryLeft`, for binding the leftmost argument of an alternating map indexed
+  by `Fin n.succ`.
+
+## Implementation notes
+`AlternatingMap` is defined in terms of `map_eq_zero_of_eq`, as this is easier to work with than
+using `map_swap` as a definition, and does not require `Neg N`.
+
+`AlternatingMap`s are provided with a coercion to `MultilinearMap`, along with a set of
+`norm_cast` lemmas that act on the algebraic structure:
+
+* `AlternatingMap.coe_add`
+* `AlternatingMap.coe_zero`
+* `AlternatingMap.coe_sub`
+* `AlternatingMap.coe_neg`
+* `AlternatingMap.coe_smul`
+-/
+
+
+-- semiring / add_comm_monoid
+
+variable {R : Type _} [Semiring R]
+
+variable {M : Type _} [AddCommMonoid M] [Module R M]
+
+variable {N : Type _} [AddCommMonoid N] [Module R N]
+
+-- semiring / add_comm_group
+
+variable {M' : Type _} [AddCommGroup M'] [Module R M']
+
+variable {N' : Type _} [AddCommGroup N'] [Module R N']
+
+variable {ι ι' ι'' : Type _}
+
+section
+
+variable (R M N ι)
+
+/-- An alternating map is a multilinear map that vanishes when two of its arguments are equal.
+-/
+structure AlternatingMap extends MultilinearMap R (fun _ : ι => M) N where
+  map_eq_zero_of_eq' : ∀ (v : ι → M) (i j : ι) (_ : v i = v j) (_ : i ≠ j), toFun v = 0
+#align alternating_map AlternatingMap
+
+end
+
+/-- The multilinear map associated to an alternating map -/
+add_decl_doc AlternatingMap.toMultilinearMap
+
+namespace AlternatingMap
+
+variable (f f' : AlternatingMap R M N ι)
+
+variable (g g₂ : AlternatingMap R M N' ι)
+
+variable (g' : AlternatingMap R M' N' ι)
+
+variable (v : ι → M) (v' : ι → M')
+
+open Function
+
+/-! Basic coercion simp lemmas, largely copied from `RingHom` and `MultilinearMap` -/
+
+
+section Coercions
+
+-- Porting note: Replaced CoeFun with FunLike instance
+instance funLike : FunLike (AlternatingMap R M N ι) (ι → M) (fun _ => N) where
+  coe f := f.toFun
+  coe_injective' := fun f g h ↦ by
+    rcases f with ⟨⟨_, _, _⟩, _⟩
+    rcases g with ⟨⟨_, _, _⟩, _⟩
+    cases h
+    rfl
```

Comment:
I think it would be reasonable to wait for https://github.com/leanprover-community/mathlib/pull/18766

Reply:
(I don't have much time this weekend; feel free to push to that PR to get it working, if you want things to move faster)